### PR TITLE
Issue #20553 - win_acl Can't set perms on "C:\Program Files" due to A…

### PR DESCRIPTION
**Pull request by JimFicarra**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

win_acl
##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 06e30485ea) last updated 2017/01/16 20:03:34 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/custom_modules']

This has existed in this module in previous 2.x releases as well.  I confirmed that this is an issue in the latest devel branch as of this writing.

```

##### SUMMARY

Fixes Issues #20553 and #15737

##### Setting permissions where ALL APPLICATION PACKAGES group applied fails (Windows 2012)

The Security.Principal.NTAccount for APPLICATION PACKAGE AUTHORITY\ALL APPLICATION PACKAGES can't be translated to its SID (Security.Principal.SecurityIdentifier).  This is an issue with the Win32 API and is reproducible using PowerShell and other tools like Sysinternals psgetsid.  This issue and its resolution (workaround) has been documented on several blogs.

- https://github.com/djberg96/win32-security/issues/6
- https://tickets.puppetlabs.com/browse/PUP-2985

This issue occurs on directories where this built-in account is applied in Windows 2012 such as the following dirs:
- c:\Program Files
- c:\Program Files (x86)
- c:\Windows

I went through Microsoft's list of Well-Known SIDS to see if there were other accounts that could not map to their SID using their fully qualified name, but this is the only one that does not map from NTAccount to SecurityIdentifier.  There were some SIDS that didn't exist in my environment so I could not test them.  The fix instantiates a new NTAccount object using the short name in the constructor so the mapping to the SID can be done.

##### Error: Set-ACL : The security identifier is not allowed to be the owner of this object.

This issue occurs when the user executing the module from the control server attempts to set permissions using win_acl on the target directory when owned by a group in which the user is not a member.

This issue is partially documented in several blogs, though I dug further and discovered it only seems to manifest itself when using Set-ACL on a dir where another group you're not in owns it.  .

- http://stackoverflow.com/questions/6622124/why-does-set-acl-on-the-drive-root-try-to-set-ownership-of-the-object
- http://www.mickputley.net/2015/11/set-acl-security-identifier-is-not.html

At first I used the suggested work-around in the code by replacing

```
$objACL = Get-ACL $path
```
with
```
$objACL = (Get-Item $Path).GetAccessControl('Access')
```

however it seems .NET 4.6.1 apparently nullifies this workaround - after the upgrade to .NET 4.6.1 I started to receive the error again.  This did not occur pre-4.6.1.  To compensate, I opted to create a function where the ACEs could be extracted from the security descriptor and used in Set-ACL.  This avoids the issue by applying only ACEs and not any owner information.

##### More Meaningful Exception Messages (not a bug)
I took the opportunity to enhance the exception handling so that any errors occurring in PowerShell are passed through the exception message. The generic messages hid the useful information underneath and was only visible when generating the module and keeping it on the remote target boxes with ANSIBLE_KEEP_REMOTE_FILES.

**_Sample Task for ALL APPLICATION_PACKAGES issue_**
```
# Create
- name: test perms
      win_acl:
        path: 'C:\Program Files\Test'
        user: Users
        rights: FullControl
        type: allow
        state: present

```
**Before:**
```
fatal: [win2012.mycompany.com]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "win_acl"
    },
    "msg": "an error occured when attempting to present FullControl permission(s) on C:\\Program Files\\Test for Users"
}
```

The existing exception handling hides the error from PowerShell, which is 
```
Exception calling "Translate" with "1" argument(s): "Some or all identity references could not be translated.
```

**After:**
```
changed: [win2012.mycompany.com] => {
    "changed": true,
    "invocation": {
        "module_name": "win_acl"
    }
}

The specified ACL is set.

```

_**Sample Task for Set-ACL issue**_
**Before:**
```
# Assign ownership of d:\win_acl_test to a new empty group before running
tasks:
    - name: test perms
      win_acl:
        path: 'd:\win_acl_test'
        user: Guests
        rights: Read
        type: allow
        state: present

#Note that for me this works from Ansible but fails in PowerShell.  Others have reported an error, and I believe this is related.

Set-ACL : The security identifier is not allowed to be the owner of this object.
At C:\Users\myuser\AppData\Local\Temp\ansible-tmp-1484854478.89-6843170110905\win_acl.ps1:404 char:13
+             Set-ACL $path $objACL
+             ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (D:\win_acl_test:String) [Set-Acl], InvalidOperationException
    + FullyQualifiedErrorId : System.InvalidOperationException,Microsoft.PowerShell.Commands.SetAclCommand

{"changed":true}
```
**After:**
```
# Copying the ACE's avoids the issue with attempting to set the owner
changed: [win2012.mycompany.com] => {
    "changed": true,
    "invocation": {
        "module_name": "win_acl"
    }
}
```

### CODE REVIEW

File: lib/ansible/modules/windows/win_acl.ps1
Comment (trondhindenes) [inline lib/ansible/modules/windows/win_acl.ps1]: I would feel better if the required cmdlets got the "ErrorAction" parameter, instead of a "blanket" ErrorActionPreference - open to feedback on that one.

Comment (JimFicarra) [reply lib/ansible/modules/windows/win_acl.ps1]: If entire blocks are being wrapped in a `try`, I'd suggest a hybrid approach; a pattern like this where we toggle it on/off as we enter the sections of code where we want to catch the exceptions.
```
If ($state -eq "present" -And $match -eq $false) {
        $ErrorActionPrference = "Stop"
        Try {
            $objACL.AddAccessRule($objACE)
            Set-ACL $path $objACL
            Set-Attr $result "changed" $true;
        }
        Catch {
            Fail-Json $result "an exception occurred when adding the specified rule - $_.Exception.Message"
        }
       finally {
           $ErrorActionPrference = "Continue"
       }
```
This is a small section of code for the example, but probably would be best to apply it at the outermost try/catch block if they are nested.

The rationale there is that if we care enough to wrap multiple statements in a try/catch block, we care enough to ensure that anything inside the block that fails should throw the exceptions.  

Limiting it to just the cmdlets opens up the possibility of letting non-terminating errors through the catch and having the module continue on and fail somewhere else and not be as helpful if the exception was caught at the point it fired.  PowerShell can skip the catch for non-terminating errors which can be undesirable.

**Comment by JimFicarra** [comment]
@trondhindenes @nitzmahone 

In an effort to address the concern of setting a blanket `ErrorActionPreference="Stop"` for the scope of this module, are there any preferred standards or conventions around Ansible's preferred usage of `erroraction` error handling with PowerShell modules when it comes to non-terminating errors?   I found 16 modules in the repo that use the blanket `ErrorActionPreference="Stop"` including win_package.  Not saying this is right or wrong - just wanting to understand the preferred standards, or at least some conventions that Ansible and the maintainers prefer.  I do think there is room for multiple error handling methods that PowerShell provides as I think they each have their use cases.  There are times when we want non-terminating errors to show and continue, be hidden, or handled explicitly.

ErrorAction/ErrorActionPreference can be overridden at more granular levels, so setting at the top isn't permanent. Plus, setting the preference explicitly at the start of a module can help with consistency across environments where the preference can be set differently in customized PowerShell startup profiles. 

I typically lean towards setting `ErrorActionPreference="Stop"` at the top and tweaking as needed within the `try/catch/finally` as in the example above, or at the cmdlet level with the `-ea` option if we want to allow non-terminating errors to not hit my handlers.  This helps reveal unanticipated non-terminating errors. (I prefer to know when it failed even if I missed a handler and it's ugly).  

Conversely, if you leave or set the default preference (`ErrorActionPreference="Continue"`) then you can also tweak the `try/catch/finally` or `-ea options` to force non-terminating errors to terminate but the difference is that non-terminating errors can slip through as a false successful run of the module.  

For additional consideration of conventions, I'd also recommend adding the actual exception message as I've done in this PR. This helps provide a bit more insight as to why the failure occurred.

I'd like to hear thoughts on this so I can make the any final changes to the PR required by the maintainers for merging.

Thanks!

**Comment by trondhindenes** [comment]
@JimFicarra I did the same over the weekend and looked at the usage of erroractionpreference. You're absolutely right - it's used in a significant number of modules (including my own), and l as you point out we don't really have coding guidelines at that level so I'll retract my statement. The important part is that the module is robust and as "serviceable" as possible.

**Comment by JimFicarra** [comment]
Thx @trondhindenes.  Could you and a few others do a final review and shipit if agreeable?
Perhaps @schwartzmx, @nitzmahone, @dagwieers

File: lib/ansible/modules/windows/win_acl.ps1
Comment (schwartzmx) [inline lib/ansible/modules/windows/win_acl.ps1]: +1 for adding these, should have been there to begin with for more detailed error info

You will need to change to `$($_.Exception.Message)` for that to expand correctly.  Or else you'll get 
```
an exception occurred when adding the specified rule - .Exception.Message
```
every time

Edit: you actually will get the upper Exception object text, with a trailing `.Exception.Message`
For example:
```powershell
pschwartz $ try { 0/0 } catch { write-host "an exception occured, $_.Exception.Message" }
an exception occured, Attempted to divide by zero..Exception.Message
```

File: lib/ansible/modules/windows/win_acl.ps1
Comment (schwartzmx) [inline lib/ansible/modules/windows/win_acl.ps1]: same here for string expansion

File: lib/ansible/modules/windows/win_acl.ps1
Comment (schwartzmx) [inline lib/ansible/modules/windows/win_acl.ps1]: same here for string expansion

File: lib/ansible/modules/windows/win_acl.ps1
Comment (schwartzmx) [inline lib/ansible/modules/windows/win_acl.ps1]: same here for string expansion

File: lib/ansible/modules/windows/win_acl.ps1
Comment (schwartzmx) [inline lib/ansible/modules/windows/win_acl.ps1]: small nitpick on no spaces between `=`

**Review by schwartzmx — COMMENTED** [review]
overall a few minor changes requested

otherwise lgtm

File: lib/ansible/modules/windows/win_acl.ps1
Comment (JimFicarra) [reply lib/ansible/modules/windows/win_acl.ps1]: Thx. I'll fix that so it evals correctly.

File: lib/ansible/modules/windows/win_acl.ps1
Comment (dagwieers) [reply lib/ansible/modules/windows/win_acl.ps1]: I prefer a default $ErrorActionPreference "Stop", and then individual -ErrorAction "Continue" for each Cmdlet that we explicitly allow to continue. At least that way it is clear what the intention was by the developer. In all other cases we may inadvertently continue on specific errors...

**Comment by JimFicarra** [comment]
Changes have been submitted, @schwartzmx

**Comment by schwartzmx** [comment]
nice work @JimFicarra 👍 

shipit

**Comment by JimFicarra** [comment]
@trondhindenes are you good here? Anything else needed to be approved for merging?

**Comment by JimFicarra** [comment]
@trondhindenes @nitzmahone @dagwieers @jborean93 I still need one more shipit - any reason why this shouldn't be merged? Hoping to get this in for 2.3

Thanks in advance.

**Comment by nitzmahone** [comment]
I'm heads-down for 2.3 core feature cutoff on 2/15- will be making a bug-scrub pass over Github issues for 2.3 not long after that, and I'll take a closer look at this during that process. Plenty of time to get this in for 2.3...

**Comment by jborean93** [comment]
@JimFicarra I will have a look at this sometime today for you if I can

**Comment by jborean93** [comment]
I haven't yet tested the code (will do so soon), but when we run the new `CopyACEs` function and remove the owner should we be setting the owner back at the end, does this actually make a change to the owner in the end?

**Comment by jborean93** [comment]
There might be an issue with this, I'm not sure if this is the correct behaviour as I haven't used win_acl before but when running
```
- name: test perms
  win_acl:
    path: C:\Program Files\Test
    user: Users
    rights: FullControl
    type: allow
    state: present
```
It does add Users with FullControl but is still inheriting the original permissions from the parent folder. It also has made a copy of each of the permissions it has inherited. I would have thought it would either remove all inheritance or just add in the new Users permissions but I could be wrong.

Before:
image

After:
image

**Comment by JimFicarra** [comment]
@jborean93 Thanks for taking the time to look at this PR.

When you apply permissions to directory in windows those get applied in conjunction with what already exists.  Inheritance of other ACEs still apply unless specifically overwritten from above and propagated downwards. Also, based on the propagation/inheritance flags that are set, it might or might not be propagated downward.  Inheritance can be disabled to remove the inherited permissions if desired.  This can be done through the UI or through some of the flags in the win_acl module.

Based on your task, your permissions are what I would expect.  Here is another example where the following conditions exist:

* d:\win_acl_test2 inherits its permissions from d:\
* The local <Server>\Users group inherits Read & Execute and Special permissions from d:\
* I apply Full Control to d:\win_acl_test2

Note that for the screenshots below I applied the permissions through the UI and not through the win-acl module.  Here are the results:
(Apologies for the blocked out device and certain usernames.  I need to mask those)

Notice that the gray checkboxes are what's inherited, and the darker checkmarks is what I applied for Users.
image

Here's what the overall permissions look like. Notice that there is a duplicate ACE for FullControl for Users.  This is the one I added, but the existing inherited permissions still are applied.
image


Effective Permissions for this folder and below are Full Control for Users.

Does this help address your concern?

**Comment by jborean93** [comment]
@JimFicarra yes I understand that it applies the changes for the Users group what I want to make sure is the correct process is that while it kept the inherited and explicity permission for Users it also made a copy of the inherited permissions for all the other groups which your screenshot doesn't show. I would have thought only the Users entry would have a duplicate (one inherited and one explicit) and the others just stay as the inherited permission.

**Comment by JimFicarra** [comment]
@jborean93 Also - about the question RE: setting the owner:

When I was troubleshooting the original issue with c:\Program Files and running locally, I ran into issues with Set-ACL, which I document as part of this PR and the issue this PR addresses as well as another issue I believe this addresses. (both linked at the start of this PR).

`Set-Acl : The security identifier is not allowed to be the owner of this object`

To summarize, however, Get-ACL retrieves the entire Security Descriptor (including the owner), and Set-ACL attempts to the ACEs and the owner. The owner issue is the rub - this is what fails in certain conditions (e.g. c:\Program Files where the folder is owned by SYSTEM and not the user making the changes). Even though it should just be setting the same owner it retrieved, it fails.

win_acl at this time does not have the capability to change the owner so I opted to remove that dependency for the time being.  A future feature for win_acl could be to change the directory or file's owner which is where the owner being set would matter.  I believe this requires a `SeRestorePrivilege` privilege to do so.

**Comment by jborean93** [comment]
Hey @JimFicarra thanks for the explanation but unfortunately I still don't know why there are duplicates for the groups we did not touch. I would have thought the normal behavior for the task I ran would be to only have an inherited and explicit entry for users and not the others? Please let me know if I am not explaining myself well.

**Comment by JimFicarra** [comment]
@jborean93 Let me attempt to duplicate your scenario and investigate.  Stay tuned...

**Comment by JimFicarra** [comment]
@jborean93 I was able duplicate what you saw.  Apologies for misunderstanding you.  Thank you for finding that issue.  

I removed the CopyACEs function - I was trying to avoid grabbing the owner in the first place because there are conditions that cause Set-ACL to fail, but the function wasn't consistent and clean.  I moved back to the Get-ACL/Set-ACL pair and to mitigate the set owner issue in specific test cases I found some methods (setting token privileges) to set the proper privileges to set the owner.

Setting the token privileges gets around some conditions I ran into during testing and it should also set us up for adding a feature to set the owner on an object as well.  

I think we also need to do a deeper dive into usage of `$objACL.AddAccessRule($objACE)`.  In one test case, it would not set replace `inherit: 'ContainerInherit, ObjectInherit'` with `inherit: None` unless I changed it to `$objACL.SetAccessRule($objACE)`.  There's also a `ResetAccessRule` method.  For now I left it as is since this is a pre-existing condition in the module.

**Comment by jborean93** [comment]
@JimFicarra that's ok, I'm not always good at explaining myself unfortunately. I've just tried testing your latest changes locally and I am getting the following error:

```
fatal: [server2016]: FAILED! => {"changed": false, "failed": true, "msg": "an error occurred when attempting to present FullControl permission(s) on C:\\Program Files\\Test for Users - Exception calling \"Translate\" with \"1\" argument(s): \"Some or all identity references could not be translated.\""}
```

tasks to replicate:
```
---
- name: create test folder
  win_file:
    path: C:\Program Files\Test
    state: directory

- name: test perms
  win_acl:
    path: C:\Program Files\Test
    user: Users
    rights: FullControl
    type: allow
    state: present
```

**Comment by JimFicarra** [comment]
@jborean93 using your tasks it succeeds for me for both 2008R2 and 2012R2. Can you provide any details about your environment that may be significant?

The error you provided is the main reason for this PR.  In the original version of this module, this line that threw this error:
```
$ruleIdentity = (New-Object Security.Principal.NTAccount 'ALL APPLICATION PACKAGES').Translate([Security.Principal.SecurityIdentifier])
```
This error occurred specifically on Windows 2012 in `c:\Program Files` and its sub-folders because the fully qualified security identifer `APPLICATION PACKAGE AUTHORITY\ALL APPLICATION PACKAGES` doesn't translate in that method (WinAPI bug/oddity).  The code I added detects this and instead translates just `ALL APPLICATION PACKAGES` to get the rule identity.   Lines 268 through 281 illustrate this handling in the [latest commit](https://github.com/Comcast/ansible/blob/e1b6169d02c8fa58ba1f0928da4a977033d94a0f/lib/ansible/modules/windows/win_acl.ps1).

As a test, I removed that section of code and duplicated your error.
```
fatal: [win2k12.mycompany.com]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "win_acl"
    },
    "msg": "an error occurred when attempting to present FullControl permission(s) on C:\\Program Files\\Test for Users - Exception calling \"Translate\" with \"1\" argument(s): \"Some or all identity references could not be translated.\""

```
Can you verify you are using the proper version of the module or confirm it's in your custom module path so that it supersedes the stock module?  Any other information that may be significant to testing this?  The only way I can duplicate this is if I remove the code that was the reason for this PR.

**Comment by jborean93** [comment]
Hey @JimFicarra sorry I was pressed for time so only tested one windows variant. I can confirm this works on Server 2008, 08 R2, 2012 and 2012 R2 but Server 2016 is failing. While we don't have integration tests running for 2016 I don't feel comfortable approving this if 2016 doesn't work.

**Comment by JimFicarra** [comment]
@jborean93 that makes sense as I did not test on 2016.  My guess is the translation of the fully qualified identifier `APPLICATION PACKAGE AUTHORITY\ALL APPLICATION PACKAGES` for  now works in 2016.  I kicked off a build this morning of 2016 from our provisioning system and will test and tweak there to handle the different behavior

**Comment by JimFicarra** [comment]
@jborean93 I've committed the appropriate fix.  So while `APPLICATION PACKAGE AUTHORITY\ALL APPLICATION PACKAGES` was introduced in Win2k12, apparently `APPLICATION PACKAGE AUTHORITY\ALL RESTRICTED APPLICATION PACKAGES` was introduced in Win2k16.  Both these identifiers are applied to `c:\Program Files and below`.  This new identifier suffered the same oddity of translation. 

I've tested on 2008/2012/2016.  Please verify at your earliest convenience. 

Incidentally, I'm not clear on the Ansible's readiness for Win2k16.  I had issues using Kerberos from Ansible against my Windows 2016 box, but was able to connect using NTLM. I'll submit a separate issue for this.

**Comment by jborean93** [comment]
Looks good, trust Microsoft to make our lives more difficult. I don't believe Ansible officially supports Win2K16 but it's bound to happen one day. I usually use ntlm or CredSSP when connecting to my boxes when testing locally as they work with both domain and local accounts.

shipit

**Comment by dagwieers** [comment]
shipit

**Comment by JimFicarra** [comment]
@nitzmahone This seems to have a conflict after it was approved for merging.  Do you need me to rebase this do a new commit or will someone else merge this?

**Comment by nitzmahone** [comment]
@JimFicarra It'd be better if you rebase it, since you know your own changes better and are less likely to trample something important... ;) Ping me when ready and I'll do final review/merge... Thanks!

**Comment by JimFicarra** [comment]
@nitzmahone the rebase is complete and merged with the reg functionality changes.

**Review by nitzmahone — APPROVED** [review]

**Comment by nitzmahone** [comment]
Thanks!

**Comment by nitzmahone** [comment]
BTW re:2016 support- we unofficially already support it, but we had some stability/flaky test issues with the AWS 2016 images when we enabled for CI, so we had to shut it off again. The modules that have tests are passing under 2016, but our CI env doesn't currently stand up a domain, so we're not doing automated tests on stuff that requires it (eg, Kerberos). I've had kerb working against 2016 before, but it's been awhile since I tried it, so it's possible that a new default security setting is breaking pywinrm's (relatively) rudimentary kerb support. CI should be re-enabled on 2016 Real Soon Now, and doing more complex automated testing with domains and non-basic auth is under discussion.

**Pull request merged**
